[Fix] login google 오류 수정

dev-msj opened the pull request:
## 작업 내용
google 성공 시 joinState를 통해 회원/비회원을 구분하여, Client에서 비회원은 회원가입 화면으로 이동하도록 수정
redirect와 queryParam을 활용하여 구글 로그인 정보를 Client의 redirect 페이지가 받아서 처리할 수 있도록 수정
Authentication 관련 exception들을 AuthenticationException으로 대체
구글 로그인 관련 Log 추가
Error Log에서 Exception Class Name을 출력하도록 수정

## 이슈 번호

dev-msj committed:
docs(application.yml) : client url 추가

dev-msj committed:
fix(OAuth2SuccessHandler) : db에 uid가 없을 때 무한 요청 이슈 해결

redirect를 통해 OAuth2 Handler 화면으로 이동할 수 있도록 수정
joinState를 통해 회원가입 여부 정보를 제공할 수 있도록 수정

dev-msj committed:
refactor(OAuth2FailureHandler) : redirect를 통해 에러 핸들링하도록 변경

dev-msj committed:
style(OAuth2SuccessHandler) : 사용하지 않는 import 제거

dev-msj committed:
style : 구글 로그인 관련 로그 추가

dev-msj committed:
feat : AuthenticationException 추가

dev-msj committed:
refactor : 인증 관련된 exception을 AuthenticationException으로 대체

dev-msj committed:
refactor : error log에서 exception class name을 표기하도록 변경

jinspark-lab commented on src/main/java/com/flab/oasis/utils/LogUtils.java:
```diff
@@ -15,11 +15,17 @@ public static void info(String message, String value) {
 
-    public static void error(ErrorCode errorCode, String message) {
-        LOGGER.error("ErrorCode: {}({})\nMessage: {}", errorCode, errorCode.getCode(), message);
+    public static void error(String exceptionClassName, ErrorCode errorCode, String message) {
```
className 을 부여하실 용도라면 다양한 방법으로 추상화가 가능할 것 같습니다. 추상화까지 너무 과하다면 Class 라도 받는게 String 보다 좋아보입니다

dev-msj replied:
네 Class 타입을 받게 하여 자신의 정보를 출력하도록 수정하였습니다!

jinspark-lab commented on src/main/resources/application.yml:
```diff
@@ -1,6 +1,9 @@
   port: 8081
 
+client:
+  url: ${CLIENT_URL}
```
이건 어떤값이 들어가는건가요?

dev-msj replied:
client server의 origin 정보를 저장합니다.
client에서 구글 로그인 결과에 해당하는 값을 받아 처리하는 페이지로 redirect 시키기 위해 추가했습니다.

jinspark-lab replied:
좋지않은 구조입니다. 이유는 한번 생각해보시면 좋을 것 같습니다. 주도권을 클라이언트에 주는걸 권고드립니다

jinspark-lab reviewed (changes requested):
피드백드렸습니다

dev-msj committed:
refactor(LogUtils) : exception을 String이 아닌 Class<?>로 받도록 수정

dev-msj committed:
chore : google oauth 관련 패키지 추가

dev-msj committed:
feat(GoogleOAuthToken) : oauth 인증 토큰을 받는 모델 추가

dev-msj committed:
feat(GoogleOAuthLoginResult) : oauth 로그인 결과를 반환하는 모델 추가

dev-msj committed:
feat(UserAuthService) : oauth 토큰를 통해 로그인하는 메서드 추가

dev-msj committed:
refactor(UserAuthController) : oauth 토큰를 통해 로그인 결과를 반환하도록 수정

dev-msj committed:
style(appliction.yml) : 사용하지 않는 값 삭제

dev-msj committed:
style : Spring Security의 Oauth 제거

dev-msj committed:
feat(LogUtils) : fatal 로그 추가

fatal 로그를 통해 stacktrace를 출력하는 기능 추가

dev-msj committed:
feat(ErrorCode) : 500 에러 추가

dev-msj committed:
refactor(UserAuthService) : fatal 로그를 남기도록 수정

dev-msj committed:
feat : stacktrace를 출력하는 FatalException 추가

dev-msj committed:
refactor : FatalException으로 대체

LogUtils를 통해 fatal 로그 출력하던 것을 FatalException을 통하도록 수정

dev-msj committed:
style(JwtFilter) : 불필요한 코드 삭제

dev-msj committed:
style(ResponseError) : 패키지 경로 수정

dev-msj committed:
test(UserAuthServiceTest) : 구글 토큰 에러 테스트 작성

잘못된 토큰이 왔을 경우 verify 메서드가 요구하는 GeneralSecurityException이나 IOException이 아닌
IllegalArgumentException을 타는 것을 확인하는 테스트

dev-msj committed:
Merge branch 'main' into fix/login-google/login-handler

jinspark-lab commented on src/test/java/com/flab/oasis/login/UserAuthServiceTest.java:
```diff
@@ -0,0 +1,27 @@
+                AuthenticationException.class,
+                () -> userAuthService.createJwtTokenByGoogleOAuthToken(googleOAuthToken)
+        );
+    }
```
기존 테스트 클래스에 합치는게 어떠실까요? UserAuth 관련 테스트 클래스가 별도로 2개가 되는데 그럴 이유가 없어보입니다

dev-msj replied:
현재 테스트 코드에서 사용 중인 Mockito에서는 테스트 대상 객체에는 InjectMock을 선언하고, DI에 필요한 객체들은 Mock을 선언하여 빈 껍데기만 주입하도록 되어 있습니다.
그런데 UserAuthControllerTest에서 UserAuthService는 Mock 객체에 해당하고, 새로 작성한 UserAuthServiceTest에서는 InjectMock에 해당해서 두 테스트가 합쳐질 수 없을 것 같습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/GoogleOAuthToken.java:
```diff
@@ -0,0 +1,14 @@
+public class GoogleOAuthToken implements Serializable {
+    private static final long serialVersionUID = -3778175011718972617L;
+
+    private String token;
```
이 부분은 클라이언트로부터 받는 google credential token 인거죠? Request / Response 용도로만 사용될 객체는 해당 네이밍을 적어주시면 좋습니다.

dev-msj replied:
네 해당내용 반영했습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/model/GoogleOAuthLoginResult.java:
```diff
@@ -0,0 +1,21 @@
+    JwtToken jwtToken;
+
+    String uid;
+    boolean joinState = false;
```
LoginResult 를 각각 따로 전달해주는 이유가 있나요? uid 랑 joinState 의 용도도 궁금합니다.

jinspark-lab commented on src/main/java/com/flab/oasis/controller/UserAuthController.java:
```diff
@@ -36,7 +38,21 @@ public boolean loginAuthFromUserLoginRequest(
+        response.addHeader(
+                CookieUtils.SET_COOKIE,
+                CookieUtils.createCookie(CookieUtils.REFRESH_TOKEN, googleOAuthLoginResult.getJwtToken().getRefreshToken())
+        );
```
UserAuthController 에서 공통 코드가 default 와 반복인데, 리팩토링 가능할 것 같습니다.

jinspark-lab reviewed (changes requested):
피드백드렸습니다.

dev-msj committed:
refactor : CookieUtils에서 setCookie 헤더를 추가하도록 수정

dev-msj committed:
refactor : 구글 로그인 요청 모델 클래스명 변경

GoogleOAuthToken -> GoogleOAuthLoginRequest

dev-msj committed:
refactor : 구글 로그인 결과 모델 클래스명 변경

GoogleOAuthLoginResult -> GoogleOAuthLoginResponse

jinspark-lab commented on src/main/java/com/flab/oasis/utils/CookieUtils.java:
```diff
@@ -18,4 +21,15 @@ public static String createCookie(String tokenType, String token) {
                 .toString();
     }
+
+    public static void setCookieHeader(HttpServletResponse response, JwtToken jwtToken) {
```
Interceptor 에 PostHandle 활용하시면 좀 더 깔끔해집니다

jinspark-lab reviewed (commented):
피드백 드렸습니다

dev-msj committed:
refactor : 구글 로그인 결과 모델명 변경

GoogleOAuthLoginResponse -> GoogleOAuthLoginResult
통신에 사용되지 않으므로 Serializable 및 JsonIgnore 삭제

dev-msj committed:
refactor(UserAuthService) : 비회원은 jwt 발급하지 않도록 수정

dev-msj committed:
refactor(UserAuthController) : set-cookie를 interceptor에서 처리하도록 수정

jwt token을 SecurityContextHolder에 등록하여 전달하도록 수정
response body는 joinState만 반환하도록 수정

dev-msj committed:
feat : jwt interceptor 추가

dev-msj committed:
test : UserAuth의 Service와 Controller Test 병합

dev-msj committed:
refactor(FatalException) : ErrorCode와 message를 받도록 변경

dev-msj committed:
refactor(UserAuthService) : if-else 구문을 간소화

dev-msj committed:
refactor : 인가 관련 ErrorCode는 403를 사용하도록 변경

dev-msj committed:
refactor(LoginResult) : 모델 수정

모델명 변경 GoogleOAuthLoginResult -> LoginResult
변수명 변경 joinState -> joinUser

dev-msj committed:
refactor(UserAuthService) : login 관련 메서드 수정

메서드명 변경
- createJwtTokenByUserLoginRequest -> tryLoginDefault
- createJwtTokenByGoogleOAuthToken -> tryLoginGoogle

기본/구글 로그인 모두 LoginResult를 반환하도록 변경

dev-msj committed:
test(UserAuthServiceTest) : UserAuthService 변경사항 반영

dev-msj committed:
refactor(UserAuthController) : LoginResult를 SecurityContext에 저장하도록 수정

dev-msj committed:
refactor(UserAuthController) : return type을 void로 변경

response 객체가 commited 상태가 되지 않도록 void로 수정
- response body가 write되면 commited 상태가 되어 response 수정이 불가능함.
- Interceptor의 postHandle에서 response 객체에 set-cookie를 하기 위함.

dev-msj committed:
refactor(JwtInterceptor) : LoginResult를 통해 쿠키를 설정하도록 수정

dev-msj committed:
refactor : JwtToken 모델명을 JsonWebToken으로 수정

JwtToken은 Json Web Token Token이므로 의미상 부적절하여 변경

dev-msj committed:
feat : UserRole Enum Class 추가

dev-msj committed:
refactor : UserAuth/Seesion에 UserRole 필드 추가

dev-msj committed:
refactor(UserAuthMapper) : select 시에 UserRole도 가져오도록 수정

dev-msj committed:
refactor(SecurityConfig) : auth 경로를 제외하고 권한을 요구하도록 수정

dev-msj committed:
refactor(JwtService) : JWT를 생성할 때 UserRole도 같이 받도록 수정

dev-msj committed:
refactor(JwtFilter) : SecurityContext에 UserRole도 저장하도록 수정

dev-msj committed:
refactor : AuthorizationException 삭제

AuthenticationException으로 대체
ErrorCode 403 삭제 후 401로 대체

dev-msj committed:
refactor(UserAuthService) : 메서드명을 getAuthenticatedUid로 수정

인가된 유저가 아닌 인증된 유저의 uid를 가져오므로 메서드명 변경

jinspark-lab commented on src/main/java/com/flab/oasis/interceptor/JwtInterceptor.java:
```diff
@@ -20,10 +19,9 @@ public void postHandle(
-        }
+        LoginResult loginResult = (LoginResult) authentication.getCredentials();
+        CookieUtils.setCookieHeader(response, loginResult.getJsonWebToken());
+        response.getOutputStream().print(loginResult.isJoinUser());
```
Interceptor 에서 응답을 조작하는 구조는 보통 좋지 않습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/UserSession.java:
```diff
@@ -16,4 +17,5 @@ public class UserSession implements Serializable {
 
     private String uid;
     private String refreshToken;
+    private UserRole userRole;
```
UserSession 에 UserRole 이 따로 있는 이유가 있나요? 설명 부탁드립니다.

dev-msj replied:
UserRole에 대한 정보가 jwt claim에 있을 경우 토큰이 탈취됐을 때 문제가 될 수 있다고 생각했습니다.
그래서 어차피 Refresh Token 검증 과정에서 uid claim을 통해 UserSession 정보를 가져와 확인하기 때문에, 여기에 UserRole을 저장하고 이 과정을 통해 UserRole을 확인하고 Security Authentication Context에 등록될 수 있도록 했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/UserAuthService.java:
```diff
@@ -103,18 +111,18 @@ private String getUidByGoogleOAuthToken(String token) {
-            if (Boolean.TRUE.equals(payload.getEmailVerified())) {
-                return payload.getEmail();
-            } else {
+            if (!Boolean.TRUE.equals(payload.getEmailVerified())) {
```
이렇게 비교하실거면 FALSE.equal 이 더 좋지않을까요?

dev-msj replied:
네 수정했습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/service/JwtService.java:
```diff
@@ -27,15 +28,13 @@ public class JwtService {
     private static final String REFRESH_TOKEN = "RefreshToken";
 
-    public JwtToken createJwtToken(String uid) {
+    public JsonWebToken createJwt(String uid, UserRole userRole) {
```
Role 발급 로직을 잘 모르겠습니다.

dev-msj replied:
UserRole은 회원가입을 통해서 발급할 예정입니다.
이 브랜치에서는 회원가입 작업을 진행하지 않기 때문에 UserRole 발급에 대한 내용이 제외되었습니다.

jinspark-lab reviewed (commented):
Role 관련 발급 로직을 잘 모르겠습니다. Authorization 은 우선순위가 높아보이지 않는데.. 이번에 추가하시려는 이유도 궁금합니다. 권한 관련 Flow 한번 설명해주시면 감사하겠습니다. 그 외에는 문제 없어보입니다.

dev-msj committed:
fix : AuthorizationException 삭제 미반영된 내용 수정

dev-msj committed:
refactor(UserAuthService) : if문 비교 조건 수정

false를 확인하는 것이 목적이므로 TRUE를 부정하는 것이 아닌 FALSE를 사용하도록 변경

dev-msj committed:
refactor(LoginResult) : uid 삭제 및 Serializable 상속

dev-msj committed:
refactor(UserAuthController) : LoginResult를 response하도록 수정

dev-msj committed:
refactor(JsonWebToken) : Serializable 상속

dev-msj committed:
feat : ResponseReissuedJWT 모델 추가

dev-msj committed:
refactor(JwtFilter) : exceptionHandle 메서드 추가

인증 오류 발생 시 해당 오류에 대한 정보를 response해줄 수 있는 메서드 추가

dev-msj committed:
style : 사용되지 않는 코드 삭제

삭제 코드
- InterceptorConfig, JwtInterceptor
- CookieUtils

jinspark-lab commented:
수고하셨습니다.
